### ISSUE TITLE
failed to 'npm run fetch'

### ISSUE THREAD

**Issue by ysun** [LLK/scratch-desktop#69]
I'm behind a corperation local network. So i have a access internet via proxy.
I can npm install successfully. But when I run 'npm run fetch' I got following error info:
```
0 info it worked if it ends with ok                                                                                            
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'run', 'fetch' ]                                                              
2 info using npm@5.6.0                                                                                                         
3 info using node@v8.11.3                                                                                                      
4 verbose run-script [ 'prefetch', 'fetch', 'postfetch' ]                                                                      
5 info lifecycle scratch-desktop@3.5.1~prefetch: scratch-desktop@3.5.1                                                         
6 info lifecycle scratch-desktop@3.5.1~fetch: scratch-desktop@3.5.1                                                            
7 verbose lifecycle scratch-desktop@3.5.1~fetch: unsafe-perm in lifecycle true                                                 
8 verbose lifecycle scratch-desktop@3.5.1~fetch: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/
works/scratch3.0/scratch-desktop/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr
/local/games:/snap/bin:/home/works/chromeos/depot_tools/:/home/works/chromeos/depot_tools/:/snap/bin:/home/works/chromeos/depot
_tools/:/home/works/chromeos/depot_tools/:/snap/bin:/snap/bin                                                                  
9 verbose lifecycle scratch-desktop@3.5.1~fetch: CWD: /home/works/scratch3.0/scratch-desktop                                   
10 silly lifecycle scratch-desktop@3.5.1~fetch: Args: [ '-c',                                                                  
10 silly lifecycle   'rimraf ./static/assets/ && mkdirp ./static/assets/ && node ./scripts/fetchMediaLibraryAssets.js' ]       
11 silly lifecycle scratch-desktop@3.5.1~fetch: Returned: code: 1  signal: null                                                
12 info lifecycle scratch-desktop@3.5.1~fetch: Failed to exec fetch script                                                     
13 verbose stack Error: scratch-desktop@3.5.1 fetch: `rimraf ./static/assets/ && mkdirp ./static/assets/ && node ./scripts/fetc
hMediaLibraryAssets.js`                                                                                                        
13 verbose stack Exit status 1                                                                                                 
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:285:16)        
13 verbose stack     at emitTwo (events.js:126:13)                                                                             
13 verbose stack     at EventEmitter.emit (events.js:214:7)                                                                    
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)     
13 verbose stack     at emitTwo (events.js:126:13)                                                                             
13 verbose stack     at ChildProcess.emit (events.js:214:7)                                                                    
13 verbose stack     at maybeClose (internal/child_process.js:925:16)                                                          
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)                                  
14 verbose pkgid scratch-desktop@3.5.1                                                                                         
15 verbose cwd /home/works/scratch3.0/scratch-desktop                                                                          
16 verbose Linux 4.15.0-34-generic                                                                                             
17 verbose argv "/usr/bin/node" "/usr/bin/npm" "run" "fetch"                                                                   
18 verbose node v8.11.3                                                                                                        
19 verbose npm  v5.6.0                                                                                                         
20 error code ELIFECYCLE                                                                                                       
21 error errno 1                                                                                                               
22 error scratch-desktop@3.5.1 fetch: `rimraf ./static/assets/ && mkdirp ./static/assets/ && node ./scripts/fetchMediaLibraryAs
sets.js`                                                                                                                       
22 error Exit status 1                                                                                                         
23 error Failed at the scratch-desktop@3.5.1 fetch script.                                                                     
23 error This is probably not a problem with npm. There is likely additional logging output above.                             
24 verbose exit [ 1, true ]                                                                                                    
```

**Comment by cwillisf** [comment]
Interesting. Could you try this command and post any error messages you see?
```
node ./scripts/fetchMediaLibraryAssets.js
```

**Comment by ysun** [comment]
Hi cwillisf,

root@pc: /home/scratch-desktop# node ./scripts/fetchMediaLibraryAssets.js
```
Total library assets: 1282

events.js:183
      throw er; // Unhandled 'error' event
      ^
Error: connect ETIMEDOUT 151.101.230.133:443
    at Object._errnoException (util.js:992:11)
    at _exceptionWithHostPort (util.js:1014:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)
```
I have to access internet via a proxy, and have setting:
```
; userconfig /root/.npmrc
https-proxy = "http://username:passwd@xxx.com:914/"
proxy = "http://username:passwd@xxx.com:914/"
registry = "https://registry.npm.taobao.org/"
strict-ssl = false
```
`npm install` and `npm run build` worked well with above config.

Do you have any alternative IP address or some other workaround solution?
It doesn't make sense why `npm fetch` needs access a server?

**Comment by ysun** [comment]
Looked into the issue.  If I deleted all code in function `fetchAsset`, the block issue could be skipped.
From that script, `ASSET_HOST = 'cdn.assets.scratch.mit.edu'`, I suspect that that IP is blocked on my side. Could you provide a non-cdn server ?

**Comment by ysun** [comment]
Hi cwillisf,
Any comments? 
So long as it skipped fetching something from server '151.101.230.133:443', the building can complete successfully.
Could you put those necessary 'MediaLibraryAssets' into the Scratch-gui repo? Or you provide an link where we can download an offline package.
I think all PRC guys can meet the same issue.

**Comment by ysun** [comment]
Hi Scrach team, any support or comments for that ?

**Comment by cwillisf** [comment]
Hmm... it might work to use this instead: `ASSET_HOST = 'assets.scratch.mit.edu'`

If that works for you, please let me know. I can at least add comments to the script for people who run into this kind of trouble.

**Comment by ysun** [comment]
Thanks for your reply. But by changing ASSET_HOST to 'assets.scratch.mit.edu', the issue is still there:

```
root@sunyi-pc:/home/works/scratch3.0/scratch-desktop# npm run fetch

> scratch-desktop@3.5.1 fetch /home/works/scratch3.0/scratch-desktop
> rimraf ./static/assets/ && mkdirp ./static/assets/ && node ./scripts/fetchMediaLibraryAssets.js

Total library assets: 1282
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: connect ETIMEDOUT 151.101.110.133:443
    at Object._errnoException (util.js:992:11)
    at _exceptionWithHostPort (util.js:1014:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! scratch-desktop@3.5.1 fetch: `rimraf ./static/assets/ && mkdirp ./static/assets/ && node ./scripts/fetchMediaLibraryAssets.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the scratch-desktop@3.5.1 fetch script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-09-03T15_03_39_499Z-debug.log

```

**Comment by cwillisf** [comment]
Unfortunately you'll need access to either `cdn.assets.scratch.mit.edu` or `assets.scratch.mit.edu` for this to work. We may eventually create a repository containing our media library assets, but doing so is not in our short-term plans.

If your network environment specifically blocks HTTPS, it might work to replace `const https = require('https');`, near the top of the file, with `const https = require('http');` instead. This isn't a change that we'd want to implement in general, but it might help in some restricted network environments.

**Comment by ysun** [comment]
Unfortunately, changing https to http is NOT helpful.
I have no idea how to workaround it, maybe I have to wait for the media lib assets repo. :(
Anyway, thanks for your info & time, Chris.

**Comment by cwillisf** [comment]
Closed in favor of #153
